network/connectionManager.go

```diff
@@ -0,0 +1,23 @@
+package network
+
+import (
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	libp2pnet "github.com/libp2p/go-libp2p-core/network"
+)
+
+type Observer interface {
+	libp2pnet.Notifiee
+}
+
+type Observable interface {
+	// indicates that the observer is ready to receive notifications from the Observable
+	Subscribe(observer Observer)
+	// indicates that the observer no longer wants to receive notifications from the Observable
+	Unsubscribe(observer Observer)
+}
```

Comment:
Could we use https://github.com/onflow/flow-go/blob/master/module/observable/observable.go here?

Reply:
so my only reservation for that is:
1. The thing being observed is an interface{} which doesn't allow strong typing and I will have to keep doing cast to to the actual type.
2. I want to support multiple types of notification (ConnectionAdded, Connection Removed, StreamAdded, StreamRemoved etc.). To accomplish that I will have to define these as separate events. To simplify I am just using libp2p Notifee as the Observer interface.

Reply:
> The thing being observed is an interface{} which doesn't allow strong typing and I will have to keep doing cast to to the actual type.

When you're writing both observable and observer, that's just a type assertion :)

> I want to support multiple types of notification (ConnectionAdded, Connection Removed, StreamAdded, StreamRemoved etc.). To accomplish that I will have to define these as separate events.

You can simply have a single event type of signature `func (libp2pnet.Notifee)`. Then your implementation of the relevant part of the Observer interface is:

```
type ConnMgrObserver struct {
 	libp2pnet.Notifiee
 }

func(o *ConnMgrObserver) OnNext(event interface{}) {
  f, ok := event.(func (libp2pnet.Notifee))
  if !ok {
     // TODO complain loudly
   }
   f(o)
}
```

The wrapper struct is just to make sure you're being specific about what you're doing. If you were to implement the `Observer` interface on say libp2pnet.Notifee directly, somebody else may want to implement that interface differently for their own purposes.